### ISSUE TITLE
Bug: [consistent-indexed-object-style] Unsafe autofix with regards to circular references [8.12.1 regression]

### ISSUE THREAD

**Issue by JavaScriptBach** [typescript-eslint/typescript-eslint#10224]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&fileType=.tsx&code=FAUwHgDg9gTgLgAjgTwiBBZZBhAljAYwFcAbAQxgBVV0BeYBBAHwQCMooSQyA7B5hDyIBbViBj8WAbwQBtAAoJcPBAGc4MZQHMAugC5MOfMXJUaCAL4BuYMAD0dhOGjwk5rHkKkK1NACYEegdGAXZObj5gxhYhUXF7R2iEACUQAlgAEwAedU0eLQAaQ08THxo-AD4rIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra1mAE0QAPRMNocARgCtEZOn0JJ0URNGgdokcGAC%2BIA0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
export type MyCircularType =
  | boolean
  | number
  | { [P in string]: MyCircularType };
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/consistent-indexed-object-style": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

This should not be a lint error. The autofix produces code that does not typecheck:

```typescript

export type MyCircularType =
  | boolean
  | number
  | Record<string, MyCircularType>;
```

`2456: Type alias 'MyCircularType' circularly references itself. 2:13 - 2:27`

This appears to be intended behavior, according to https://github.com/microsoft/TypeScript/issues/35164

### Actual Result

Autofixed into code that doesn't typecheck:

```typescript

export type MyCircularType =
  | boolean
  | number
  | Record<string, MyCircularType>;
```

### Additional Info

Probably caused by https://github.com/typescript-eslint/typescript-eslint/pull/10160.

**Comment by JavaScriptBach** [comment]
Took a cursory look at this. It seems like there is an already an attempt to have this rule not fail if there are circular index signatures, but it doesn't handle the following cases:

1) The `in` operator
```typescript
type MyCircularType2 = { [P in "foo" | "bar"]: MyCircularType2 };
```

2) Circular references between 2 different type aliases
```typescript
type Bar = { [k: string]: Foo };
type Foo = string | Bar;
```

I'm guessing that fixing 1) is tractable, but I'm not sure about 2).

@kirkwaiblinger I can try to write a PR to fix one or both if you have some pointers? Is https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/consistent-indexed-object-style.ts#L70-L75 the place to fix?

**Comment by bradzacher** [comment]
It's doubtful that we can handle the second case.
But the first case should be handleable and should already be handled by the logic.

**Comment by kirkwaiblinger** [comment]
Yeah, case 1 was for sure an oversight. I simply didn't think of possibly-circular types in #10160. Seems reasonable to check the circularity detection logic you've linked for the `TSIndexSignature` case, and try to port it to the `TSMappedType` case.

tbh Case 2 feels like it could be detectable too if you want to wanted to implement a cycle detection algorithm for the type variables. Questionable how worthwhile this is, perhaps. But, conceptually I'd think we could analyze syntactically:
- `type Foo`'s declaration depends on `Bar` and string
- `Bar's` declaration is visible in this file. So we'll check what types its declaration depends on
- `type Bar`'s declaration depends on `Foo` and string
- we have found a cycle.
(note that I'm spitballing; I haven't tried doing something like this and don't know of an example in this repo offhand)

**Comment by kirkwaiblinger** [comment]
> @kirkwaiblinger I can try to write a PR to fix one or both if you have some pointers? Is https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/consistent-indexed-object-style.ts#L70-L75 the place to fix?

Following from previous comment, I think that the issue for 1 is that 

https://github.com/typescript-eslint/typescript-eslint/blob/9c956ee1ca1330c312ee51bf26c01c281b7665e0/packages/eslint-plugin/src/rules/consistent-indexed-object-style.ts#L146-L208

simply doesn't hit any of the logic in 

https://github.com/typescript-eslint/typescript-eslint/blob/9c956ee1ca1330c312ee51bf26c01c281b7665e0/packages/eslint-plugin/src/rules/consistent-indexed-object-style.ts#L67-L79

So I'd guess that copying/porting/unifying the logic would address point 1.

**Comment by kirkwaiblinger** [comment]
Note that the "in operator" here isn't really an "in operator". It's just the syntax that distinguishes a mapped type (`type T = { [K in Foo]: SomeType<K> }`) from an index signature in a type literal (`type T = { [k: string]: number }`). They look similar but have vastly capabilities (I learned while working on this 🤣 ).

Mentioning this clarification in case it helps make sense of the current code referring to `TSMappedType`.
